src/index.ts

```diff
@@ -92,3 +92,102 @@ export function render(
     })
     .catch(error => showError(el, error));
 }
+
+interface WidgetUpdate {
+  key: string;
+  remove?: string;
+  insert?: any[];
+}
+
+interface WidgetUpdateMessage {
+  type: "update";
+  updates: WidgetUpdate[];
+}
+
+// validate the ev object and cast it to the correct type
+function checkWidgetUpdate(ev: any): WidgetUpdateMessage | null {
+  if (ev.type != "update") {
+    return null;
+  }
+
+  // TODO: Fully validate ev and give a easy to understand error message if it is ill-formed
+  return ev as WidgetUpdateMessage;
+}
+
+// NOTE: juggle to support optional dependencies that don't break webpack and jupyter
+var VegaWidgetDef: any = null;
+if (__webpack_modules__[require.resolveWeak("@jupyter-widgets/base")]) {
+  const widgets = require("@jupyter-widgets/base");
+
+  VegaWidgetDef = widgets.DOMWidgetView.extend({
+    render: function() {
+      this.viewElement = document.createElement("div");
+      this.el.appendChild(this.viewElement);
+
+      this.errorElement = document.createElement("div");
+      this.errorElement.style.color = "red";
+      this.el.appendChild(this.errorElement);
+
+      const reembed = () => {
+        if (this.view) {
+          this.view.finalize();
+          this.view = null;
+        }
+        const spec = JSON.parse(this.model.get("_spec_source"));
+
+        if (spec == null) {
+          return;
+        }
+
+        vegaEmbed(this.viewElement, spec)
```

Comment:
Add support for EmbedOptions as above.

Reply:
We should probably also set the loader as above.

You can ignore the mode since you are setting it via the spec. However, I'd say that the docs need to be updated to explain how to switch from VL to Vega.

Reply:
I added the EmbedOptions in the same way as the spec is handled, i.e., they can be set at construction and dynamically. I also added a corresponding section to the example notebook (and discovered / fixed a small bug with dynamic spec updates).

Regarding the mode. TBH, I was kind of unaware of the mode option, just copied the spec from the existing example notebooks and was happy it worked. 

If you like, I can add the same split into `VegaWidget` and `VegaLiteWidget` as for the non-widget code parts. Via subclassing the changes should be minimal.

Reply:
Nah, I think relying on $schema is enough, I think.

Reply:
I added a sentence about the ``$schema`` property to the docstring.